ships/0031-shipwright-trigger.md

```diff
@@ -0,0 +1,296 @@
+   	 },
+   	 "id": "commit-id",
+   	 "timestamp": "2022-02-18T11:02:02.635313+01:00"
+    }
+}
+```
+
+Note the snippet attribute `.repository.htm_url` can be compared to Build's `.spec.source.url` in order to determine which `Build` objects are affected by the event. Likewise, we can employ `.ref` to determine which branch is affected, and later asseting if it's part of the Build `.spec.trigger.when[].branches` entries.
+
+Depending on the event type, it will need to inspect `.head_commit.id` as well, so in other words, the events offer more than enough information to fulfill the needs of triggering Shipwright Builds (3).
+
+### Git Service Providers Support
+
+Shipwright Trigger will initially ship with [GitHub support](https://github.com/google/go-github), users will be able to configure the WebHook endpoint with additional [secret-token verification](https://docs.github.com/en/developers/webhooks-and-events/webhooks/securing-your-webhooks#setting-your-secret-token), in order to trigger Shipwright Builds accordingly.
+
+The GitHub support is based on a common interface, on which we will support more service providers in the near future, like for instance GitLab and others. Therefore, the interface semantics must be able to implement the following use cases:
+
+- Parse the WebHook request payload identifying the specific type of event
+- Extract the inputs needed to be able to select the Build instances
+- Validate the payload against a secret-token shared between the service provider and the user
+
+With that said, the [initial interface suggestion](https://github.com/otaviof/shipwright-trigger/blob/main/pkg/trigger/webhooks/interface.go) is as follows:
+
+
+```go
+// Interface describe the signature expected for the instances handling WebHook
+// requests, coming from Git service providers.
+type Interface interface {
+    // ExtractRequestPayload parse and extract the request details, is service 
+    // provider specific.
+    ExtractRequestPayload(*http.Request) (*RequestPayload, error)
+
+    // ExtractBuildSelector extract the search parameters to select the Build 
+    // objects related to the WebHook request payload.
+    ExtractBuildSelector(*RequestPayload) (*BuildSelector, error)
+
+    // ValidateSignature verifies the request payload against the informed
+    // secret token.
+    ValidateSignature(*RequestPayload, []byte) error
+}
+```
+
+Internally, we will differentiate the service provider based on the [pattern](https://pkg.go.dev/net/http#Handle), the request URL path, and therefore we can allow users to configure which pattern will respond for a given Git service provider requests.
+
+### Security Model
+
+The security model is enforced by only allowing the Shipwright Trigger to have the [least amount of permissions as possible](https://github.com/otaviof/shipwright-trigger/blob/main/deploy/200-role.yaml). The new application will have to watch over Shipwright Build, BuildRun, Tekton PipelineRun and Run.
```

Comment:
Missing `Image` here ?

Reply:
We might as well say all the resources involved with the Shipwright Trigger controllers. And for the record, I still need to prototype the workflow with the `Image` resource.